Update built in functions. (#140)
storyscript/docs.storyscript.io: storyscript/functions/README.md

# Functions

Repetitive code blocks can be grouped into functions for better re-usability and modularity:

```coffeescript
function get_user id: int returns Map[string, any]
    someone = (psql exec query: "select * from users where id={id} limit 1;")[0]
    someone["contact"] = fullcontact person email: someone["email"]
    return someone as Map[string, any]

user_a = get_user(id: 7)
user_b = get_user(id: 10)
```

The example above is a function that queries the database and also retrieves their `FullContact` profile.

Functions must define their inputs and their output which help with transparency, auto-complete and type checking.

Functions can have only one return value. If a function declares a return value, then that function **MUST** use `return` to return a value of that type.

```coffeescript
function add a: int b: int returns int
   return a + b
```

Functions that do not have an output may omit the `returns` keyword in it's declaration.

> ## Invoking multiple functions on the same line 

Author: -1 for nested - nothing is nested here. `on the same line` conveys the effect more clearly.

Repeating code blocks can be grouped into functions for better reusability and modularity:

```coffeescript

> function get_user id: int returns Map[string, any]

Author: None really, just wanted to introduce whitespaces for readability.

> function get_user id: int returns Map[string, any]

Author: Good point, I like the one with whitespaces :) (inspired from Objective C arguments).

> function get_user id: int returns Map[string, any]

Author: It really depends on the eyes to be honest - purely subjective.

> function get_user id: int returns Map[string, any]

Author: Settled down on "x:x" instead of "x: x" (offline poll).

```coffeescript
arr = [1, 2, 3, 4, 5]
arr.length()  # returns the length of the list, 5 in this case
arr.append(item:6)  # returns a new list with the item `6` added to the end of it
arr.prepend(item:1)  # returns a new list, with the item `1` added to the start of it
arr.random()  # returns a random element from this list
arr.reverse()  # returns a new list, in reverse order
arr.sort()  # returns a new list, sorted in an ascending fashion
arr.min()  # returns the lowest of the items in this list (if it contains numbers)
arr.max()  # returns the largest of the items in this list (if it contains numbers)
arr.sum()  # returns the sum of all the items in this list (if it contains numbers)
arr.contains(item:3)  # returns true if 3 is present, false otherwise
arr.unique()  # returns a new list, which will contain only unique items
arr.remove(item:3)  # returns a new list, with the specified item removed from the list
arr.index(of:5)  # returns the index of an item, 4 in this case
arr.replace(item:3 by:4)  # returns a new list, where all occurrences of the item `3` is replaced with `4`
arr.slice(start:1)  # returns a new list, with items starting from the index `1` (inclusive)

> arr.slice(start:1 end:3)  # returns a new list, with items starting from the index `1` (inclusive), to the index `3` (inclusive) 

Reviewer: We should mention somewhere an example or at least state that indices start from 0.

Author: https://github.com/storyscript/docs.storyscript.io/issues/143

Author: Created https://github.com/storyscript/docs.storyscript.io/issues/143

    return 1
>>> Error: syntax error in story at line 5, column 12

> 5|               return 1

Author: Ah that's okay - it's in the error, not in any code.

> 5|               return 1

Reviewer: No it's not, sorry. This will result in a misalignment.

> 5|               return 1

Author: Looks like I ditched this. No need for an error here.

                 ^

E0110: Function has no return output defined. Only `return` is allowed.
```

## Invoking multiple functions on the same line 

```coffeescript

> result = functionA(arg: functionB(arg: functionC()))

Author: Why can't they be the same? Introducing numbering makes the arguments appear to be related.

> result = functionA(arg: functionB(arg: functionC()))

Reviewer: See below. It allows to reference to them in the text for a better explanation.

str.length()  # returns the number of UTF-8 characters
str.replace(item: "ab" by: "AB")  # returns a string by replacing "ab" with "AB"
str.replace(pattern: /ab/ by: "AB")  # replace all occurrences of the pattern RegExp /ab/ with "ab"
str.contains(pattern: /a/)  # returns true if the string matches the pattern, false otherwise
str.contains(item: "hello")  # returns true if the string contains the sequence "hello", false otherwise
str.split(by: ".")  # returns a list by splitting the string with the delimiter
str.uppercase()  # returns a string where all characters are upper cased
str.lowercase()  # returns a string where all characters are lower cased
str.capitalize()  # returns a string where the first letter of each word is capitalized (eg: "jane smith" becomes "Jane Smith")
str.trim() # returns a string with any leading and trailing whitespace (including tabs) removed
str.startswith(prefix: "abc") # returns true if the string starts with the prefix "abc"
str.endswith(suffix: "xyz") # returns true if the string ends with the suffix "xyz"

> "hey there".substring(start: 4)  # returns "there" 

Author: Not the fourth character, but rather the index. In the string example, the fourth character is " ".

> "hey there".substring(start: 2 end: 4)  # returns "y t"

Author: Let's keep the talk in indexes instead of characters?

num.tan()  # returns the tangent of a number
num.asin()  # returns the arcsine of a number
num.acos()  # returns the arccosine of a number
num.atan()  # returns the arctangent of a number
num.log()  # returns the logarithmic of a number
num.log2()  # returns the base 2 logarithm of a number
num.log10()  # returns the base 10 logarithm of a number
num.exp()  # returns e^num, where e is Euler's number
num.abs()  # returns the absolute value of a number
num.isNaN()  # returns true if the number is not a valid number, false otherwise
num.isInfinity()  # returns true if the number represents infinity, false otherwise
num.approxEqual(value: 10.50001)  # returns true if the number is approximately equal to 10.50001
num.approxEqual(value: 10.50001 maxRelDiff: 1.5)  # returns true if the number is approximately equal to 10.50001, within a tolerance 
num.approxEqual(value: 10.50001 maxAbsDiff: 1.5)  # TODO - @wilzbach
num.approxEqual(value: 10.50001 maxRelDiff: 1.5 maxAbsDiff: 1.5)  # TODO - @wilzbach
num.sqrt()  # returns the square root of a number
```

### Lists

```coffeescript
arr = [1, 2, 3, 4, 5]
arr.length()  # returns the length of the list, 5 in this case

> arr.append(item: 6)  # returns a new list, with the item `6` added to the end of it

Reviewer: ```suggestion
arr.append(item: 6)  # returns a new list with the item `6` added to the end of it
```

why the comma?

arr.prepend(item: 1)  # returns a new list, with the item `1` added to the start of it
arr.random()  # returns a random element from this list
arr.reverse()  # returns a new list, in reverse order

> arr.sort()  # returns the list, sorted in an ascending fashion

Reviewer: ```suggestion
arr.sort()  # returns a new list, sorted in an ascending fashion
```

arr.min()  # returns the lowest of the items in this list (if it contains numbers)
arr.max()  # returns the largest of the items in this list (if it contains numbers)
arr.sum()  # returns the sum of all the items in this list (if it contains numbers)
arr.contains(item: 3)  # returns true if 3 is present, false otherwise
arr.unique()  # returns a new list, which will contain only unique items
arr.remove(item: 3)  # returns a new list, with the specified item removed from the list
arr.index(of: 5)  # returns the index of an item, 4 in this case
arr.replace(item: 3 by: 4)  # returns a new list, where all occurrences of the item `3` is replaced with `4`
arr.slice(start: 1)  # returns a new list, with items starting from the index `1` (inclusive)

> arr.slice(start: 1 end: 3)  # returns a new list, with items starting from the index `1` (inclusive), to the index `3` (inclusive) 

Reviewer: ```suggestion
arr.slice(start: 1 end: 3)  # returns a new list, with items starting from the index `1` (inclusive), to the index `3` (exclusive) 
```

Author: Let's make both inclusive? Otherwise it's a little tricky to remember which one is inclusive, and which one is exclusive.

m.length()  # returns the size of the map, 2 in this case
m.keys()  # returns a list of all keys
m.values()  # returns a list of all values

> m.remove(key: "c")  # returns a map with the key 'c' in it

Reviewer: ```suggestion
m.remove(key: "c")  # returns a map without the key "c" in it
```

> m.flatten()  # returns a list of key/value pairs (eg: [['a', 1], ['b', 2]])

Reviewer: ```suggestion
m.flatten()  # returns a list of key/value pairs (eg: [["a", 1], ["b", 2]])
```

Don't use single quotes here. Storyscript doesn't support them.

> m.contains(key: "c")  # returns true if the key 'c' exists in the map, false otherwise

Reviewer: ```suggestion
m.contains(key: "c")  # returns true if the key "c" exists in the map, false otherwise
```

> m.contains(value: "c")  # returns true if the value 'c' exists in the map, false otherwise

Reviewer: ```suggestion
m.contains(value: "c")  # returns true if the value "c" exists in the map, false otherwise
```

> m.get(key: "b" default: "." )  # returns the value for the key 'b', or a 'default' value if it doesn't exist

Reviewer: ```suggestion
m.get(key: "b" default: "." )  # returns the value for the key "b", or a default value if it doesn't exist
```
